---
title: "scNMTseq challenge analysis using a PLS-based approach"
author: Al JalalAbadi^[Melbourne Integrative Genomics, The University of Melbourne (al.jal.abadi@gmail.com)]
date: "`r format(Sys.Date(), '%d %b %Y')`"
output: rmarkdown::html_vignette
params:
  ## setup params
  on_my_mac: !r Sys.info()['user'] == 'alabadi'
  save_output: false
  local_data: false
  mini_run: false
  matching_rna_for_umap: false
  ## run params
  drop_lineages: !r c('Primitive_endoderm','Visceral_endoderm', 'ExE_ectoderm')
  umap_params: !r c(run.seed = 42, n_neighbors = 15, n_components = 2, min_dist = 0.55)
  ggplot_alpha: 0.8
  pls_ncomp:  4
vignette: >
  %\VignetteIndexEntry{scNMTseq challenge analysis using a PLS-based approach}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


--------

```{r, include = FALSE}
library(knitr)
knitr::opts_chunk$set(collapse = TRUE, 
                      comment = "#>", 
                      fig.width = 8, 
                      cache = FALSE)
```

**Note:** These analyses are different from the ones presented in the hackathon due to updated preprocessing if data in order to harmonize the analyses for publication. For original analyses see https://github.com/ajabadi/scNMT_seq_gastrulation.
  
--------

Load the required packages:

```{r, eval=params$on_my_mac, include=FALSE}
sapply(list.files('../R', full.names = TRUE), source)
local.lib <- '../lib'
dir.create(local.lib)
.libPaths(local.lib)
## uncomment this if you are not building the vignette along with the package
# remotes::install_github('mixOmicsTeam/mixOmics@MultiAssayExperiment', upgrade = 'never')
```

```{r load packages, warning=FALSE, message=FALSE}
library(BIRSBIO2020.scNMTseq.PLS)
library(MultiAssayExperiment)
library(scater)
library(scran)
library(mixOmics)
library(ggplot2)
library(magrittr)
library(reshape2)
library(uwot)
```

```{r, eval=params$save_output, include=params$save_output}
## create directories to save the figures and output data:
if (!dir.exists('figures')) {
  cat('creating "figures" folder ...\n')
  dir.create('figures')
}

if (!dir.exists('savedata')) {
   cat('creating "savedata" folder ...\n')
  dir.create('savedata')
}
```

# Data

Details of the hackathon data and preprocessing steps: https://github.com/BIRSBiointegration/Hackathon/tree/master/scNMT-seq

Load the `MultiAssayExperiment` object from Cloudstor:
```{r, eval=!params$mini_run}
cat('loading data from Cloudstor ...\n')
gastru.mae_path <- url('https://cloudstor.aarnet.edu.au/plus/s/jsW7nh4YwThw8Q5/download')
```

```{r, eval=params$on_my_mac & !params$mini_run, include=FALSE}
cat('loading data from local folder ...\n')
gastru.mae_path <- 'savedata/scnmtseq_gastrulation_mae-sce.rds'
```

```{r, eval=params$on_my_mac & params$mini_run, include=FALSE}
cat('loading mini data from local folder ...\n')
gastru.mae_path <- 'savedata/mini.gastru.mae.rds'
```

```{r load data}
gastru.mae <- readRDS(gastru.mae_path)
```

<!-- ```{r, eval=!params$on_my_mac} -->
<!-- cat(sprintf('loading the following assays using SingleCellMultiModal package: %s\n', paste(params$keep_assays, collapse = ', '))) -->
<!-- gastru.mae <- scNMT(dataType = 'mouse_gastrulation', modes = params$keep_assays, dry.run = FALSE, verbose = FALSE) -->
<!-- ``` -->

## Filter MAE object

```{r}
## subset RNA expression and DNA methylation modalities
keep_assays <- grep("rna|met",names(assays(gastru.mae)))
gastru.mae <- gastru.mae[,,keep_assays]
## remove putative extraembryonic cells
cat(sprintf('dropping lineages %s, plus the unassigned lineages.\n', paste(params$drop_lineages, collapse = ', ')))
gastru.mae <- gastru.mae[,!(gastru.mae$lineage %in% params$drop_lineages | is.na(gastru.mae$lineage)),]
## keep cells which pass RNA QC
gastru.mae <- gastru.mae[,gastru.mae$pass_rnaQC==TRUE,]
## Keep full rna SCE for UMAP
rna.sce <- gastru.mae@ExperimentList$rna
## keep cells that also pass QC for DNA methylation
gastru.mae <- gastru.mae[,gastru.mae$pass_metQC==TRUE,]
## rna SCE for cells passing met and rna QC
rna.sce.matching <- gastru.mae@ExperimentList$rna
```

Replace the rna SCE with logcounts in MAE object as the integration wrapper requires matrices in assays:

```{r}
## rna is an SCE object
gastru.mae@ExperimentList
## replace SCE with logcounts
gastru.mae@ExperimentList$rna <- logcounts(gastru.mae@ExperimentList$rna)
## rna updated to matrix of logcounts
gastru.mae@ExperimentList
```

Overview of assays:

```{r}
## helper to get % of missing values in an array
get_pct_missing <- function(arr) {
  round(100*sum(is.na(arr))/prod(dim(arr)))
}
## get dimensions and % of missing values for all assays
df <- lapply(experiments(gastru.mae), function(w)
     {
     c(
       'N (# cells)' = dim(w)[2],
       'P (# features)' = dim(w)[1],
       '% data missing' = get_pct_missing(w)
     )
     })
df <- data.frame(df)
kable(t(df), align = 'l')
```

Breakdown of the number of cells in each stage:

```{r, warning=FALSE}
table(gastru.mae$stage) %>% as.data.frame() %>% t() %>% 
  set_rownames(c('stage', '# of cells')) %>% kable()
```

## Feature detection

Create density plots of the feature detection rate across all cells for all modalities:

```{r, message=FALSE}
# get the methylation assays
met_assays <- grep(names(gastru.mae), pattern = '^met', value = TRUE)
# add dimensions to labels for ggplot
dims <- lapply(experiments(gastru.mae[,,met_assays]), dim)
dims <- sapply(dims, function(x) sprintf(' (%s, %s)', x[2], x[1]))
names(met_assays) <- paste0(met_assays, dims) %>% as.list()
# calculate the feature detection in a data.frame for methylation assays
coverages <- lapply(met_assays, function(assay_name) {
  mat <- assay(gastru.mae, assay_name)
  NAs <- rowSums(!is.na(mat))/dim(mat)[2]*100
  data.frame(pct_NAs=NAs)
})
# create a long data.frame containing the assay name for plot
coverages <- rbindListWithNames(coverages)
coverages$dataset <- factor(coverages$dataset, levels = unique(coverages$dataset), ordered = TRUE)
```

Genomic contexts vary in level of feature detection rates:

```{r, fig.width=8, fig.asp=0.4}
cov_plot <- ggplot(coverages, aes(x = pct_NAs)) + geom_density(fill = 'lightblue', show.legend = FALSE) +
  geom_vline(aes(xintercept=mean(pct_NAs)),
             color="blue", linetype="dashed", size=0.5) +
  labs(x = '% of cells detecting the feature') + facet_wrap(.~dataset, nrow = 2) +
  theme_bw() + theme(strip.text.x = element_text(size = 10, face = 'bold', color = 'purple'))
  
cov_plot
```

Density plots for methylation data show that shorter genomic regions tend to have less feature coverage. Dashed blue line indicates the average across all modalities.

```{r, eval=params$save_output, include=params$save_output}
ggsave(cov_plot, filename = 'figures/covplots.pdf', width = 8, height = 4)
```

# RNA

```{r}
# Create colour palettes for stages:
stages <- c("E4.5", "E5.5", "E6.5", "E7.5")
# col_palette <- dput(viridisLite::viridis(n = length(stages)))
col_palette <- c("#440154FF", "#31688EFF", "#35B779FF", "#FDE725FF")
names(col_palette) <- stages
```

```{r}
## helper function to create UMAP plots coloured by stage
plot_reducedDims <- function(sce, reducedDim = 'UMAP', 
                             dims = c(1,2), col_palette) {
  comps.prefix <- ifelse(reducedDim == 'UMAP', 'UMAP_', 'PC_')
  df <- data.frame(reducedDim(sce, reducedDim))[,dims]
  axes <- colnames(df) <- paste0(comps.prefix, dims)
  df$stage <- sce$stage
  ggplot(df, aes_string(axes[1], axes[2])) + geom_point(aes(col=stage), alpha = params$ggplot_alpha) +
    theme_classic()+ scale_color_manual(values = col_palette) + labs(col = 'Stage')
}
```

## Dimension reduction all of RNA cells

Select highly variable genes after accounting for technical variation:

```{r}
decomp <- modelGeneVar(rna.sce)
## filter by mean expression and significance of biological variation signal
hvgs <- rownames(decomp)[decomp$p.value<0.01 & decomp$mean > 0.01]
length(hvgs)
```

Run PCA and then UMAP using PCs:

```{r}
npcs <- 15
## PCA first: retrieve npcs PCs
rna.sce <- runPCA(rna.sce,  ncomponents = npcs, subset_row=hvgs, name = 'PCA')

## UMAP parameters used:
cat(sprintf('Running UMAP with parameters %s\n', paste(names(params$umap_params), ':',params$umap_params, collapse = ', ')))

## run UMAP
set.seed(params$umap_params['run.seed'])
rna.sce <- runUMAP(rna.sce, dimred="PCA", 
                   ncomponents = params$umap_params['n_components'], 
                   n_neighbors = params$umap_params['n_neighbors'], 
                   min_dist = params$umap_params['min_dist'])
```

### UMAP

Plot the first two UMAP components:

```{r, fig.asp=0.7}
plot_reducedDims(sce = rna.sce, reducedDim = 'UMAP', dims = c(1,2), col_palette = col_palette)
```

UMAP plot separates the cell populations from early-stage, mid-stage, and late-stage cells and captures the the within-stage variation for early-stage and mid-stage cells.

## Dimension reduction of all matching RNA cells used in integration

Select highly variable genes after accounting for technical variation:

```{r}
decomp <- modelGeneVar(rna.sce.matching)
## filter by mean expression and significance of biological variation signal
hvgs <- rownames(decomp)[decomp$p.value<0.01 & decomp$mean > 0.01]
length(hvgs)
```

Run PCA and then UMAP using PCs:

```{r}
npcs <- 15
## PCA first: retrieve npcs PCs
rna.sce.matching <- runPCA(rna.sce.matching, ncomponents = npcs, subset_row=hvgs, name = 'PCA')

## UMAP parameters used:
cat(sprintf('Running UMAP with parameters %s\n', paste(names(params$umap_params), ':',params$umap_params, collapse = ', ')))

## run UMAP
set.seed(params$umap_params['run.seed'])
rna.sce.matching <- runUMAP(rna.sce.matching, dimred="PCA", 
                   ncomponents = params$umap_params['n_components'], 
                   n_neighbors = params$umap_params['n_neighbors'], 
                   min_dist = params$umap_params['min_dist'])
```

### PCA

Plot the first two UMAP components:

```{r, fig.asp=0.7}
plot_reducedDims(sce = rna.sce.matching, reducedDim = 'PCA', dims = c(1,2), col_palette = col_palette)
```


### UMAP

Plot the first two UMAP components:

```{r, fig.asp=0.7}
plot_reducedDims(sce = rna.sce.matching, reducedDim = 'UMAP', dims = c(1,2), col_palette = col_palette)
```

Similar results to the analysis with all cells.

# Integration

<!-- ## gene id to gene symbol conversion data.frame -->

<!-- Here we create a reference data.frame to convert Ensembl gene ids to gene symbols for GSEA. -->

<!-- Get all the gene names from all modalities: -->

<!-- ```{r} -->
<!-- all_features <- sapply(experiments(gastru.mae), rownames) %>% unlist() -->
<!-- all_genes <- grep(pattern = '^ENSMUSG', x = all_features, value = TRUE) -->
<!-- all_genes <- unique(all_genes) -->
<!-- length(all_genes) -->
<!-- ``` -->

<!-- Get the Ensembl gene name for these gene ids: -->

<!-- ```{r} -->
<!-- ensembl <- useMart("ensembl", dataset="mmusculus_gene_ensembl") -->
<!-- mouse_gene_ids  <- all_genes -->

<!-- symbols <- getBM(attributes=c('ensembl_gene_id', -->
<!--                               'external_gene_name'), -->
<!--                  filters = "ensembl_gene_id", -->
<!--                  values = all_genes, -->
<!--                  mart = ensembl) -->
<!-- symbols <- data.frame(symbols, row.names = 1) -->
<!-- ``` -->

<!-- ```{r, eval=params$save_output, echo=FALSE} -->
<!-- save(symbols, file = 'savedata/all-gene-symbols.RData') -->
<!-- # load('savedata/all-gene-symbols.RData') -->
<!-- ``` -->

## PLS

Data dimensions overview:

```{r}
# data dimensions
lapply(experiments(gastru.mae), dim)
```

I   ntegration of all modalities with feature selection:

<!-- ```{r, eval=params$save_output, echo=params$save_output} -->
<!-- pls_output_name <- sprintf('multimodal_spls-ncomp_%s-scale_.rds', ncomp, scale) -->
<!-- ``` -->

Here we select for 50 features on each component in `multimodal_sPLS_wrapper`:

```{r, echo=FALSE}
## helper function to format runtime
format_runtime <- function(dsec) {
  hours <- floor(dsec / 3600)
  minutes <- floor((dsec - 3600 * hours) / 60)
  seconds <- dsec - 3600*hours - 60*minutes
  paste0(
    sapply(c(hours, minutes, seconds), function(x) {
      formatC(x, width = 2, format = "d", flag = "0")
    }), collapse = ":")
}
```

```{r}
## number of components
ncomp <- params$pls_ncomp
## feature scaling
scale <- FALSE
```

```{r run pls, warning=FALSE}
cat(sprintf('Running PLS with %s components performing variable selection\n', ncomp))
st <- system.time({
  mmspls <-
    multimodal_sPLS_wrapper(mae = gastru.mae, study_assays = NULL,
                            ncomp = ncomp, scale = scale, design = 'null', lineages = NULL,
                            stages = NULL, DA = NULL, keepX = NULL, save = FALSE)
})['elapsed']

cat('\nPLS run finished. Runtime: ', format_runtime(st), '\n')
```

```{r, eval=params$save_output, echo=params$save_output}
saveRDS(mmspls, file = 'savedata/MultiModalSparsePLS-All.rds')
# mmspls <- readRDS('savedata/MultiModalSparsePLS-All.rds')
```

```{r}
plot_pls <- function(pls_obj, stage, legend.title = 'Stage', comp = c(1,2), col_palette) {
  variates <- lapply(pls_obj$variates, function(arr){
    df <- as.data.frame(arr[,comp])
    df$stage <- stage
    df
  })
  variates <- rbindListWithNames(variates, new_col = 'Modality')
  axes <- colnames(variates)[1:2]
  ggplot(variates) + 
    theme_classic() +
    geom_point(aes_string(x = axes[1], y = axes[2], col = 'stage'), alpha = params$ggplot_alph) +
    facet_wrap(.~Modality, ncol = 2) +
    scale_color_manual(values = col_palette)
}
```

## score plots

Colored by stage:

```{r, fig.width=8, fig.height=8}
plot_pls(pls_obj = mmspls, stage = gastru.mae$stage, col_palette = col_palette)
```

The derived latent components which seek to maximise the sum of concordant variation b/w rna and other modalities show phenotypic relevance in terms of the embryonic stages. It indicates coordinated changes across the methylomes associated with the distinct transcriptional states of cells in the regions of study during gastrulation.

## UMAP using `r ncomp` PLS components

Add PLS embeddings to SCE so UMAP is applied using `scater::runUMAP`:

```{r}
for (block in names(mmspls$variates)) {
  variates <- mmspls$variates[block]
  dimred.name <- paste0('PLS_', block)
  umap.name <- paste0('UMAP_PLS_', block)
  reducedDim(rna.sce.matching, dimred.name) <- mmspls$variates[[block]]
  set.seed(params$umap_params['run.seed'])
  rna.sce.matching <- runUMAP(rna.sce.matching, dimred=dimred.name, 
                              ncomponents = params$umap_params['n_components'], 
                              n_neighbors = params$umap_params['n_neighbors'], 
                              min_dist = params$umap_params['min_dist'], 
                              name = umap.name)
}
```

Create a long data.frame of all variates for facet plotting: 
```{r}
all_pls_umap <- list()
for (block in names(mmspls$variates)) {
  umap.name <- paste0('UMAP_PLS_', block)
  umap_df <- as.data.frame(reducedDim(rna.sce.matching, umap.name))
  colnames(umap_df) <- paste0('UMAP_', seq_along(umap_df))
  umap_df$stage <- gastru.mae$stage
  all_pls_umap[[block]] <- umap_df
}
all_pls_umap <- rbindListWithNames(all_pls_umap, new_col = 'Modality')
```

Plot UMAP:

```{r}
axes <- colnames(all_pls_umap)[1:2]
  ggplot(all_pls_umap) + 
    theme_classic() +
    geom_point(aes_string(x = axes[1], y = axes[2], col = 'stage'), alpha = params$ggplot_alph) +
    facet_wrap(.~Modality, ncol = 2) +
    scale_color_manual(values = col_palette)
```

UMAP projection using `r ncomp` PLS components for each modality shows the distinct states of early stage cells on transcriptional and DNA methylation level. These states are coordinated across both molecular levels. The projection also demonstrates the transition from mid-stage to late-stage for single cells across different views.

```{r, eval=params$save_output, echo=FALSE}
ggsave(filename = 'figures/plotIndiv-MultiModalSparsePLS-stage.pdf', width = 8, height = 6)
```

# concordance with RNA

Here we calculate the correlations between components from gene expression and different methylome modalities:

```{r}
## For a pls object which includes pls cell embeddings from a computed latent space
## where covariance with rna is maximal, calculate the correlation b/w rna embeddings
## and those of a given modality's (per component and average) as a measure of concordant variation with rna
concordance <- function(pls_obj, comp = 1) {
  
  ## correlation of components with RNA components
  cor_with_y <- sapply(pls_obj$variates[-1], function(x_variates) {
    y_variates <- pls_obj$variates[[1]]
    y_variates <- y_variates[, comp, drop = FALSE]
    cor <- cor(y_variates, x_variates)
    cor <- diag(cor)
    names(cor) <- paste0('Component ', comp)
    cor['mean'] <- mean(cor)
    cor <- round(cor, 2)
    cor
  })
  print(kable(cor_with_y))
  ## make a long data.frame
  cor_with_y <- cor_with_y %>% as.data.frame() %>% t() %>% 
    reshape2::melt(id.vars = 1:3) %>% 
    set_colnames(c('Modality', 'Component', 'Concordance'))
  ## print concordance measures for each componet and the average:
  # sapply(unique(as.character(cor_with_y$Component)), function(z){
  #   cat('#> Component:', z )
  #   dplyr::filter(cor_with_y, Component == z)[,c('Modality', 'Concordance')] %>% 
  #     kable() 
  # })

  ggplot(cor_with_y, aes(x = Modality, y = Concordance)) + geom_col(aes(fill = Component), position = 'dodge2') + theme_classic() +
    labs(y = sprintf('Correlation with RNA (comp %s) ', paste(comp, collapse = ', ')), x = '', fill = '') + ylim(c(0, 1)) + theme(axis.text.x = element_text(angle = 45, vjust = 0.8, hjust=1))
}
```

```{r}
concordance(pls_obj = mmspls, comp = c(1, 2))
```

# Conclusion

Multi-modal PLS derives the common variation between transcriptome and methylome in different genomic regions for `r dim(gastru.mae@ExperimentList$rna)[2]` single cells during mouse gastrulation. The measurements include high levels of masked values and the method ignores these missing elements during integration. The common axes of variation are mainly driven by the stages of embryo and there is high coordination between transcriptional and CpG methylation states.
