---
title: "scNMTseq challenge analysis using a PLS-based approach"
author: Al JalalAbadi^[Melbourne Integrative Genomics, The University of Melbourne (al.jal.abadi@gmail.com)]
date: "`r format(Sys.Date(), '%d %b %Y')`"
output: rmarkdown::html_vignette
params:
  ## setup params
  on_my_mac: !r Sys.info()['user'] == 'alabadi'
  save_output: false
  local_data: false
  mini_run: false
  matching_rna_for_umap: false
  ## run params
  drop_lineages: !r c('Primitive_endoderm','Visceral_endoderm', 'ExE_ectoderm')
  umap_params: !r c(run.seed = 42, n_neighbors = 15, n_components = 2, min_dist = 0.55)
  ggplot_alpha: 0.8
  pls_ncomp:  4
vignette: >
  %\VignetteIndexEntry{scNMTseq challenge analysis using a PLS-based approach}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


--------

```{r, include = FALSE}
library(knitr)
knitr::opts_chunk$set(collapse = TRUE, 
                      comment = "#>", 
                      fig.width = 8, 
                      cache = FALSE)
```

**Note:** These analyses are different from the ones presented in the hackathon due to updated preprocessing if data in order to harmonize the analyses for publication. For original analyses see https://github.com/ajabadi/scNMT_seq_gastrulation.
  
--------

Load the required packages:

```{r, eval=params$on_my_mac, include=FALSE}
sapply(list.files('../R', full.names = TRUE), source)
local.lib <- '../lib'
dir.create(local.lib)
.libPaths(local.lib)
## uncomment this if you are not building the vignette along with the package
# remotes::install_github('mixOmicsTeam/mixOmics@MultiAssayExperiment', upgrade = 'never')
```

```{r load packages, warning=FALSE, message=FALSE}
library(BIRSBIO2020.scNMTseq.PLS)
library(MultiAssayExperiment)
library(scater)
library(scran)
library(mixOmics)
library(ggplot2)
library(magrittr)
library(reshape2)
library(uwot)
library(impute)
```

```{r, eval=params$save_output, include=params$save_output}
## create directories to save the figures and output data:
if (!dir.exists('figures')) {
  cat('creating "figures" folder ...\n')
  dir.create('figures')
}

if (!dir.exists('savedata')) {
   cat('creating "savedata" folder ...\n')
  dir.create('savedata')
}
```

# Data

Details of the hackathon data and preprocessing steps: https://github.com/BIRSBiointegration/Hackathon/tree/master/scNMT-seq

Load the `MultiAssayExperiment` object from Cloudstor:
```{r, eval=!params$mini_run}
cat('loading data from Cloudstor ...\n')
gastru.mae_path <- url('https://cloudstor.aarnet.edu.au/plus/s/jsW7nh4YwThw8Q5/download')
```

```{r, eval=params$on_my_mac & !params$mini_run, include=FALSE}
cat('loading data from local folder ...\n')
gastru.mae_path <- 'savedata/scnmtseq_gastrulation_mae-sce.rds'
```

```{r, eval=params$on_my_mac & params$mini_run, include=FALSE}
cat('loading mini data from local folder ...\n')
gastru.mae_path <- 'savedata/mini.gastru.mae.rds'
```

```{r load data}
gastru.mae <- readRDS(gastru.mae_path)
```

<!-- ```{r, eval=!params$on_my_mac} -->
<!-- cat(sprintf('loading the following assays using SingleCellMultiModal package: %s\n', paste(params$keep_assays, collapse = ', '))) -->
<!-- gastru.mae <- scNMT(dataType = 'mouse_gastrulation', modes = params$keep_assays, dry.run = FALSE, verbose = FALSE) -->
<!-- ``` -->

## Filter MAE object

Keep cells which pass QC metrics and exclude putative extra-embryonic cells as they are starkly different in methylation patterns and will drive most of variation in the relatively small population of matching cells, not allowing to fully explore cell to cell variation intended to explore in this hackathon:

```{r}
## subset RNA expression and DNA methylation modalities
keep_assays <- grep("rna|met",names(assays(gastru.mae)))
gastru.mae <- gastru.mae[,,keep_assays]
## remove putative extraembryonic cells
cat(sprintf('dropping lineages %s, plus the unassigned lineages.\n', paste(params$drop_lineages, collapse = ', ')))
gastru.mae <- gastru.mae[,!(gastru.mae$lineage %in% params$drop_lineages | is.na(gastru.mae$lineage)),]
## keep cells which pass RNA QC
gastru.mae <- gastru.mae[,gastru.mae$pass_rnaQC==TRUE,]
## Keep full rna SCE for UMAP
rna.sce <- gastru.mae@ExperimentList$rna
## keep cells that also pass QC for DNA methylation
gastru.mae <- gastru.mae[,gastru.mae$pass_metQC==TRUE,]
## rna SCE for cells passing met and rna QC
rna.sce.matching <- gastru.mae@ExperimentList$rna
```

Replace the rna SCE with normalised counts in MAE object as the integration wrapper requires matrices in assays:

```{r}
## rna is an SCE object
gastru.mae@ExperimentList
## replace SCE with logcounts
gastru.mae@ExperimentList$rna <- logcounts(gastru.mae@ExperimentList$rna)
## rna updated to matrix of logcounts
gastru.mae@ExperimentList
```

Overview of assays:

```{r}
## helper to get % of missing values in an array
get_pct_missing <- function(arr) {
  round(100*sum(is.na(arr))/prod(dim(arr)))
}
## get dimensions and % of missing values for all assays
df <- lapply(experiments(gastru.mae), function(w)
     {
     c(
       'N (# cells)' = dim(w)[2],
       'P (# features)' = dim(w)[1],
       '% data missing' = get_pct_missing(w)
     )
     })
df <- data.frame(df)
kable(t(df), align = 'l')
```

Breakdown of the number of cells in each stage:

```{r, warning=FALSE}
table(gastru.mae$stage) %>% as.data.frame() %>% t() %>% 
  set_rownames(c('stage', '# of cells')) %>% kable()
```

## Feature detection

Create density plots of the feature detection rate across all cells for all modalities:

```{r, message=FALSE, echo=FALSE}
# get the methylation assays
met_assays <- grep(names(gastru.mae), pattern = '^met', value = TRUE)
# add dimensions to labels for ggplot
dims <- lapply(experiments(gastru.mae[,,met_assays]), dim)
dims <- sapply(dims, function(x) sprintf(' (%s, %s)', x[2], x[1]))
names(met_assays) <- paste0(met_assays, dims) %>% as.list()
# calculate the feature detection in a data.frame for methylation assays
coverages <- lapply(met_assays, function(assay_name) {
  mat <- assay(gastru.mae, assay_name)
  NAs <- rowSums(!is.na(mat))/dim(mat)[2]*100
  data.frame(pct_NAs=NAs)
})
# create a long data.frame containing the assay name for plot
coverages <- rbindListWithNames(coverages)
coverages$dataset <- factor(coverages$dataset, levels = unique(coverages$dataset), ordered = TRUE)
```

```{r, fig.width=8, fig.asp=0.4}
cov_plot <- ggplot(coverages, aes(x = pct_NAs)) + geom_density(fill = 'lightblue', show.legend = FALSE) +
  geom_vline(aes(xintercept=mean(pct_NAs)),
             color="blue", linetype="dashed", size=0.5) +
  labs(x = '% of cells detecting the feature') + facet_wrap(.~dataset, nrow = 2) +
  theme_bw() + theme(strip.text.x = element_text(size = 10, face = 'bold', color = 'purple'))
  
cov_plot
```

Density plots for methylation data show that shorter genomic regions tend to have less feature coverage. Dashed blue line indicates the average across all modalities.

```{r, eval=params$save_output, include=params$save_output}
ggsave(cov_plot, filename = 'figures/covplots.pdf', width = 8, height = 4)
```

# RNA

```{r}
# Create colour palettes for stages:
stages <- c("E4.5", "E5.5", "E6.5", "E7.5")
# col_palette <- dput(viridisLite::viridis(n = length(stages)))
col_palette <- c("#440154FF", "#31688EFF", "#35B779FF", "#FDE725FF")
names(col_palette) <- stages
```

```{r}
## helper function to create UMAP plots coloured by stage
plot_reducedDims <- function(sce, reducedDim = 'UMAP', 
                             dims = c(1,2), col_palette) {
  comps.prefix <- ifelse(reducedDim == 'UMAP', 'UMAP_', 'PC_')
  df <- data.frame(reducedDim(sce, reducedDim))[,dims]
  axes <- colnames(df) <- paste0(comps.prefix, dims)
  df$stage <- sce$stage
  ggplot(df, aes_string(axes[1], axes[2])) + geom_point(aes(col=stage), alpha = params$ggplot_alpha) +
    theme_classic()+ scale_color_manual(values = col_palette) + labs(col = 'Stage')
}
```

## Dimension reduction all of RNA cells

Select highly variable genes after accounting for technical variation:

```{r}
decomp <- modelGeneVar(rna.sce)
## filter by mean expression and significance of biological variation signal
hvgs <- rownames(decomp)[decomp$p.value<0.01 & decomp$mean > 0.01]
length(hvgs)
```

Run PCA and then UMAP using PCs:

```{r}
npcs <- 15
## PCA first: retrieve npcs PCs
rna.sce <- runPCA(rna.sce,  ncomponents = npcs, subset_row=hvgs, name = 'PCA')

## UMAP parameters used:
cat(sprintf('Running UMAP with parameters %s\n', paste(names(params$umap_params), ':',params$umap_params, collapse = ', ')))

## run UMAP
set.seed(params$umap_params['run.seed'])
rna.sce <- runUMAP(rna.sce, dimred="PCA", 
                   ncomponents = params$umap_params['n_components'], 
                   n_neighbors = params$umap_params['n_neighbors'], 
                   min_dist = params$umap_params['min_dist'])
```

### UMAP

Plot the first two UMAP components:

```{r, fig.asp=0.7}
plot_reducedDims(sce = rna.sce, reducedDim = 'UMAP', dims = c(1,2), col_palette = col_palette)
```

UMAP plot separates the cell populations from early-stage, mid-stage, and late-stage cells and captures the the within-stage variation for early-stage and mid-stage cells.

## Dimension reduction of all matching RNA cells used in integration

Same procedure as before with the matching cells:
```{r}
decomp <- modelGeneVar(rna.sce.matching)
## filter by mean expression and significance of biological variation signal
hvgs <- rownames(decomp)[decomp$p.value<0.01 & decomp$mean > 0.01]
length(hvgs)

npcs <- 15
## PCA first: retrieve npcs PCs
rna.sce.matching <- runPCA(rna.sce.matching, ncomponents = npcs, subset_row=hvgs, name = 'PCA')

## UMAP parameters used:
cat(sprintf('Running UMAP with parameters %s\n', paste(names(params$umap_params), ':',params$umap_params, collapse = ', ')))

## run UMAP
set.seed(params$umap_params['run.seed'])
rna.sce.matching <- runUMAP(rna.sce.matching, dimred="PCA", 
                   ncomponents = params$umap_params['n_components'], 
                   n_neighbors = params$umap_params['n_neighbors'], 
                   min_dist = params$umap_params['min_dist'])
```

### Dimension reduction

PCA:

```{r, fig.asp=0.7}
plot_reducedDims(sce = rna.sce.matching, reducedDim = 'PCA', dims = c(1,2), col_palette = col_palette)
```


UMAP:

```{r, fig.asp=0.7}
plot_reducedDims(sce = rna.sce.matching, reducedDim = 'UMAP', dims = c(1,2), col_palette = col_palette)
```

Similar results to the analysis with all cells.

# Integration

<!-- ## gene id to gene symbol conversion data.frame -->

<!-- Here we create a reference data.frame to convert Ensembl gene ids to gene symbols for GSEA. -->

<!-- Get all the gene names from all modalities: -->

<!-- ```{r} -->
<!-- all_features <- sapply(experiments(gastru.mae), rownames) %>% unlist() -->
<!-- all_genes <- grep(pattern = '^ENSMUSG', x = all_features, value = TRUE) -->
<!-- all_genes <- unique(all_genes) -->
<!-- length(all_genes) -->
<!-- ``` -->

<!-- Get the Ensembl gene name for these gene ids: -->

<!-- ```{r} -->
<!-- ensembl <- useMart("ensembl", dataset="mmusculus_gene_ensembl") -->
<!-- mouse_gene_ids  <- all_genes -->

<!-- symbols <- getBM(attributes=c('ensembl_gene_id', -->
<!--                               'external_gene_name'), -->
<!--                  filters = "ensembl_gene_id", -->
<!--                  values = all_genes, -->
<!--                  mart = ensembl) -->
<!-- symbols <- data.frame(symbols, row.names = 1) -->
<!-- ``` -->

<!-- ```{r, eval=params$save_output, echo=FALSE} -->
<!-- save(symbols, file = 'savedata/all-gene-symbols.RData') -->
<!-- # load('savedata/all-gene-symbols.RData') -->
<!-- ``` -->

## PLS

Integration of all modalities with feature selection.

Here we select for 50 features on each component in `multimodal_sPLS_wrapper`:

```{r, echo=FALSE}
## helper function to format runtime
format_runtime <- function(dsec) {
  hours <- floor(dsec / 3600)
  minutes <- floor((dsec - 3600 * hours) / 60)
  seconds <- dsec - 3600*hours - 60*minutes
  paste0(
    sapply(c(hours, minutes, seconds), function(x) {
      formatC(x, width = 2, format = "d", flag = "0")
    }), collapse = ":")
}
```

```{r}
## number of components
ncomp <- params$pls_ncomp
## feature scaling
scale <- FALSE
```

```{r run pls, warning=FALSE}
cat(sprintf('Running PLS with %s components performing variable selection\n', ncomp))
st <- system.time({
  mmspls <-
    multimodal_sPLS_wrapper(mae = gastru.mae, study_assays = NULL,
                            ncomp = ncomp, scale = scale, design = 'null', lineages = NULL,
                            stages = NULL, DA = NULL, keepX = NULL, save = FALSE)
})['elapsed']
mmspls$stage <- gastru.mae$stage
mmspls$lineage<- gastru.mae$lineage
cat('\nPLS run finished. Runtime: ', format_runtime(st), '\n')
```

```{r, eval=params$save_output, echo=params$save_output}
saveRDS(mmspls, file = 'savedata/MultiModalSparsePLS-All.rds')
# mmspls <- readRDS('savedata/MultiModalSparsePLS-All.rds')
```

```{r, echo=FALSE}
plot_pls <- function(pls_obj, stage, legend.title = 'Stage', comp = c(1,2), col_palette) {
  variates <- lapply(pls_obj$variates, function(arr){
    df <- as.data.frame(arr[,comp])
    df$stage <- stage
    df
  })
  variates <- rbindListWithNames(variates, new_col = 'Modality')
  axes <- colnames(variates)[1:2]
  axes_labels <- paste0('component ', comp)
  ggplot(variates) + 
    theme_classic() +
    geom_point(aes_string(x = axes[1], y = axes[2], col = 'stage'), alpha = params$ggplot_alph) +
    facet_wrap(.~Modality, ncol = 2, scales = "free") +
    scale_color_manual(values = col_palette) +
    labs(x = axes_labels[1], y = axes_labels[2], col = legend.title)
}
```

## score plots

Plot PLS components for cells across all views colored by embryonic stages:

```{r, fig.width=8, fig.height=8}
pls_no.impute <- plot_pls(pls_obj = mmspls, stage = mmspls$stage, col_palette = col_palette)
pls_no.impute
```
```{r, eval=params$save_output, echo=FALSE}
ggsave(filename = 'figures/plotIndiv-MultiModalSparsePLS-stage.pdf')
```

The derived latent components which seek to maximise the sum of concordant variation between rna and other modalities separate early-stage cells from late-stage cells. It indicates coordinated changes across the methylomes associated with the distinct transcriptional states of cells in the regions of study during gastrulation. 

Coloring the cells by putative lineages mapped using rna cluster signature genes highlights the variation within the late-stage cells driven by disparate cell subpopulations in the transcriptome view:

```{r, fig.width=8, fig.height=8}
plot_pls(mmspls, mmspls$lineage, 
         col_palette = gg_color_hue(length(unique(gastru.mae$lineage))), 
         legend.title = 'Lineage')
```

<!-- ## UMAP using `r ncomp` PLS components -->

<!-- Add PLS embeddings to SCE so UMAP is applied using `scater::runUMAP`: -->

<!-- ```{r} -->
<!-- for (block in names(mmspls$variates)) { -->
<!--   variates <- mmspls$variates[block] -->
<!--   dimred.name <- paste0('PLS_', block) -->
<!--   umap.name <- paste0('UMAP_PLS_', block) -->
<!--   reducedDim(rna.sce.matching, dimred.name) <- mmspls$variates[[block]] -->
<!--   set.seed(params$umap_params['run.seed']) -->
<!--   rna.sce.matching <- runUMAP(rna.sce.matching, dimred=dimred.name,  -->
<!--                               ncomponents = params$umap_params['n_components'],  -->
<!--                               n_neighbors = params$umap_params['n_neighbors'],  -->
<!--                               min_dist = params$umap_params['min_dist'],  -->
<!--                               name = umap.name) -->
<!-- } -->
<!-- ``` -->

<!-- Create a long data.frame of all variates for facet plotting:  -->
<!-- ```{r} -->
<!-- all_pls_umap <- list() -->
<!-- for (block in names(mmspls$variates)) { -->
<!--   umap.name <- paste0('UMAP_PLS_', block) -->
<!--   umap_df <- as.data.frame(reducedDim(rna.sce.matching, umap.name)) -->
<!--   colnames(umap_df) <- paste0('UMAP_', seq_along(umap_df)) -->
<!--   umap_df$stage <- gastru.mae$stage -->
<!--   umap_df$lineage <- gastru.mae$lineage -->
<!--   all_pls_umap[[block]] <- umap_df -->
<!-- } -->
<!-- all_pls_umap <- rbindListWithNames(all_pls_umap, new_col = 'Modality') -->
<!-- ``` -->

<!-- Plot UMAP: -->

<!-- ```{r} -->
<!-- ## ------------- stage -->
<!-- axes <- colnames(all_pls_umap)[1:2] -->
<!-- gg_stage <- ggplot(all_pls_umap) +  -->
<!--   theme_classic() + -->
<!--   geom_point(aes_string(x = axes[1], y = axes[2], col = 'stage'), alpha = params$ggplot_alph) + -->
<!--   facet_wrap(.~Modality, ncol = 2, scales = 'free') + -->
<!--   scale_color_manual(values = col_palette) + -->
<!--   labs(col = 'Stage') -->

<!-- ## ------------- lineage -->
<!-- # lineage_col_palette <- viridisLite::viridis(n=length(unique(unique(gastru.mae$lineage)))) -->
<!-- lineage_col_palette <- c("#440154FF", "#3B528BFF", "#21908CFF", "#5DC863FF", "#FDE725FF") -->
<!-- names(lineage_col_palette) <- unique(gastru.mae$lineage) -->
<!-- axes <- colnames(all_pls_umap)[1:2] -->
<!-- gg_lineage <- ggplot(all_pls_umap) +  -->
<!--   theme_classic() + -->
<!--   geom_point(aes_string(x = axes[1], y = axes[2], col = 'lineage'), alpha = params$ggplot_alph) + -->
<!--   facet_wrap(.~Modality, ncol = 2, scales = 'free') + -->
<!--   scale_color_manual(values = lineage_col_palette) + -->
<!--   labs(col = 'Putative Lineage') -->
<!-- ``` -->

<!-- ```{r, fig.width=12, fig.height=6} -->
<!-- gg_stage_lineage <- gridExtra::grid.arrange(gg_stage + labs(x = ''), -->
<!--                         ggplot() + theme_minimal() + labs(x = 'UMAP_1'),  -->
<!--                         gg_lineage + labs(y = '', x = ''),  -->
<!--                         widths = c(10,1,10),  -->
<!--                         nrow=1) -->
<!-- ``` -->


<!-- UMAP projection using `r ncomp` PLS components for each modality shows the distinct states cells from each stage on transcriptional and DNA methylation level in the projected space. This suggests that the molecular changes during mouse gastrulation are coordinated across the study modalities. The projection also highlights the different patterns of transcriptional variation among different putative lineages in late-stage cells. -->

<!-- ```{r, eval=params$save_output, echo=FALSE} -->
<!-- ggsave(plot = gg_stage_lineage, filename = 'figures/UMAP-MultiModalSparsePLS-stage-lineage.pdf', width = 12, height = 6) -->
<!-- ggsave(plot = gg_stage, filename = 'figures/UMAP-MultiModalSparsePLS-stage.pdf', width = 10, height = 9) -->
<!-- ``` -->

<!-- ## concordance with RNA -->

<!-- Here we calculate the correlations between components from gene expression and different methylome modalities as a measure of concordnat variation: -->

<!-- ```{r} -->
<!-- ## For a pls object which includes pls cell embeddings from a computed latent space -->
<!-- ## where covariance with rna is maximal, calculate the correlation b/w rna scores -->
<!-- ## and those of a given modality's (per component and average) - potentially weighted by amount of  -->
<!-- ## variance each rna component explains (weighted=TRUE) - as a measure of concordant variation with rna -->
<!-- concordance <- function(pls_obj,  -->
<!--                         comp = 1, ## which componets to use -->
<!--                         weighted = TRUE,  -->
<!--                         which.plot = NULL, ## which component to plot (or [weighted.]mean) -->
<!--                         show.legend = TRUE -->
<!--                         ) { -->
<!--   if (!is.null(which.plot)) { -->
<!--     which.plot[is.numeric(which.plot)] <- paste0('Component ',  which.plot[is.numeric(which.plot)]) -->
<!--   } -->
<!--   ## correlation of components with RNA components -->
<!--   cor_with_y <- sapply(pls_obj$variates[-1], function(x_variates) { -->
<!--     y_variates <- pls_obj$variates[[1]] -->
<!--     y_variates <- y_variates[, comp, drop = FALSE] -->
<!--     cor <- cor(y_variates, x_variates) -->
<!--     cor <- diag(cor) -->
<!--     names(cor) <- paste0('Component ', comp) -->

<!--     if (isFALSE(weighted)) { -->
<!--       weight <- rep(1, length(cor)) -->
<!--     } else { -->
<!--       weight <- pls_obj$explained_variance$rna -->
<!--     } -->
<!--     cor.mean <- mean(cor) -->
<!--     wtd.cor.mean <- weighted.mean(x = cor, w = weight) -->
<!--     cor['mean'] <- cor.mean -->
<!--     cor['weighted.mean'] <- wtd.cor.mean -->
<!--     cor <- round(cor, 2) -->
<!--     cor -->
<!--   }) -->
<!--   print(kable(cor_with_y)) -->
<!--   ## make a long data.frame -->
<!--   cor_with_y <- cor_with_y %>% as.data.frame() %>% t() %>%  -->
<!--     reshape2::melt(id.vars = 1:3) %>%  -->
<!--     set_colnames(c('Modality', 'Component', 'Concordance')) -->

<!--   if (!is.null(which.plot)) { -->
<!--     cor_with_y <- cor_with_y[cor_with_y$Component %in% which.plot,] -->
<!--   } -->
<!--   ggplot(cor_with_y, aes(x = Modality, y = Concordance)) + geom_col(aes(fill = Component), position = 'dodge2', show.legend = show.legend) + theme_classic() + -->
<!--     labs(y = sprintf('Correlation with RNA (comp %s) ', paste(comp, collapse = ', ')), x = '', fill = '') + ylim(c(0, 1)) + theme(axis.text.x = element_text(angle = 45, vjust = 0.8, hjust=1)) -->
<!-- } -->
<!-- ``` -->

<!-- ```{r} -->
<!-- concordance(pls_obj = mmspls, comp = c(1:4), which.plot = NULL) -->
<!-- ``` -->

# Imputation

Impute methylation data:

```{r, results='hide', message=FALSE}
# sum(!gastru.mae$pass_metQC)
#> 0
gastru.mae.imputed <- gastru.mae
met_assays <- grep(pattern = '^met', names(experiments(gastru.mae.imputed)), value = TRUE)
for (assay_type in met_assays) {
  assay_values <- gastru.mae.imputed@ExperimentList[[assay_type]]
  cat('\nimputing values for:', assay_type, '\n')
  gastru.mae.imputed@ExperimentList[[assay_type]] <- 
    impute.knn(data = assay_values, 
               k = 30, 
               rowmax = 0.99, 
               colmax = 0.99,  
               rng.seed=42)$data
}
```
```{r, eval=params$save_output, echo=params$save_output}
saveRDS(gastru.mae.imputed, file = 'savedata/gastru.mae.imputed.rds')
# gastru.mae.imputed <- readRDS('savedata/gastru.mae.imputed.rds')
```

Perform data integration using imputed methylation data:
```{r run pls imputed, warning=FALSE}
cat(sprintf('Running PLS on imputed data with %s components performing variable selection\n', ncomp))
st <- system.time({
  mmspls.imputed <-
    multimodal_sPLS_wrapper(mae = gastru.mae.imputed, study_assays = NULL,
                            ncomp = ncomp, scale = scale, design = 'null', lineages = NULL,
                            stages = NULL, DA = NULL, keepX = NULL, save = FALSE)
})['elapsed']

cat('\nPLS run finished. Runtime: ', format_runtime(st), '\n')
```

```{r, eval=params$save_output, echo=FALSE}
ggsave(filename = 'figures/plotIndiv-MultiModalSparsePLS-stage.pdf')
```

```{r, eval=params$save_output, echo=params$save_output}
saveRDS(mmspls.imputed, file = 'savedata/MultiModalSparsePLS-All-imputed.rds')
# mmspls.imputed <- readRDS('savedata/MultiModalSparsePLS-All.rds')
```

## score plots

Plot PLS components for cells across all views:
```{r, fig.width=8, fig.height=8}
pls_impute <- plot_pls(pls_obj = mmspls.imputed, stage = gastru.mae.imputed$stage, col_palette = col_palette)
pls_impute
```
```{r, eval=params$save_output, echo=FALSE}
ggsave(filename = 'figures/plotIndiv-MultiModalSparsePLS-imputed-stage.pdf')
```

<!-- ## UMAP using `r ncomp` PLS components from imputed data -->

<!-- ```{r} -->
<!-- for (block in names(mmspls.imputed$variates)) { -->
<!--   variates <- mmspls.imputed$variates[block] -->
<!--   dimred.name <- paste0('PLS_imputed_', block) -->
<!--   umap.name <- paste0('UMAP_PLS_imputed_', block) -->
<!--   reducedDim(rna.sce.matching, dimred.name) <- mmspls.imputed$variates[[block]] -->
<!--   set.seed(params$umap_params['run.seed']) -->
<!--   rna.sce.matching <- runUMAP(rna.sce.matching, dimred=dimred.name,  -->
<!--                               ncomponents = params$umap_params['n_components'],  -->
<!--                               n_neighbors = params$umap_params['n_neighbors'],  -->
<!--                               min_dist = params$umap_params['min_dist'],  -->
<!--                               name = umap.name) -->
<!-- } -->
<!-- ``` -->

<!-- Create a long data.frame of all variates for facet plotting:  -->
<!-- ```{r} -->
<!-- all_pls_umap <- list() -->
<!-- for (block in names(mmspls.imputed$variates)) { -->
<!--   umap.name <- paste0('UMAP_PLS_imputed_', block) -->
<!--   umap_df <- as.data.frame(reducedDim(rna.sce.matching, umap.name)) -->
<!--   colnames(umap_df) <- paste0('UMAP_', seq_along(umap_df)) -->
<!--   umap_df$stage <- gastru.mae.imputed$stage -->
<!--   umap_df$lineage <- gastru.mae.imputed$lineage -->
<!--   all_pls_umap[[block]] <- umap_df -->
<!-- } -->
<!-- all_pls_umap <- rbindListWithNames(all_pls_umap, new_col = 'Modality') -->
<!-- ``` -->

<!-- Plot UMAP: -->

<!-- ```{r} -->
<!-- ## ------------- stage -->
<!-- axes <- colnames(all_pls_umap)[1:2] -->
<!-- gg_stage <- ggplot(all_pls_umap) +  -->
<!--   theme_classic() + -->
<!--   geom_point(aes_string(x = axes[1], y = axes[2], col = 'stage'), alpha = params$ggplot_alph) + -->
<!--   facet_wrap(.~Modality, ncol = 2, scales = 'free') + -->
<!--   scale_color_manual(values = col_palette) + -->
<!--   labs(col = 'Stage') -->

<!-- ## ------------- lineage -->
<!-- # lineage_col_palette <- viridisLite::viridis(n=length(unique(unique(gastru.mae.imputed$lineage)))) -->
<!-- lineage_col_palette <- c("#440154FF", "#3B528BFF", "#21908CFF", "#5DC863FF", "#FDE725FF") -->
<!-- names(lineage_col_palette) <- unique(gastru.mae.imputed$lineage) -->
<!-- axes <- colnames(all_pls_umap)[1:2] -->
<!-- gg_lineage <- ggplot(all_pls_umap) +  -->
<!--   theme_classic() + -->
<!--   geom_point(aes_string(x = axes[1], y = axes[2], col = 'lineage'), alpha = params$ggplot_alph) + -->
<!--   facet_wrap(.~Modality, ncol = 2, scales = 'free') + -->
<!--   scale_color_manual(values = lineage_col_palette) + -->
<!--   labs(col = 'Putative Lineage') -->
<!-- ``` -->

<!-- ```{r, fig.width=12, fig.height=6} -->
<!-- gg_stage_lineage <- gridExtra::grid.arrange(gg_stage + labs(x = ''), -->
<!--                         ggplot() + theme_minimal() + labs(x = 'UMAP_1'),  -->
<!--                         gg_lineage + labs(y = '', x = ''),  -->
<!--                         widths = c(10,1,10),  -->
<!--                         nrow=1) -->
<!-- ``` -->

<!-- UMAP projection using imputed data: -->

<!-- ```{r, eval=params$save_output, echo=FALSE} -->
<!-- ggsave(plot = gg_stage_lineage, filename = 'figures/UMAP-MultiModalSparsePLS-imputed-stage-lineage.pdf', width = 12, height = 6) -->
<!-- ggsave(plot = gg_stage, filename = 'figures/UMAP-MultiModalSparsePLS-imputed-stage.pdf', width = 10, height = 9) -->
<!-- ``` -->

## Predictive performance of PLS components

We finally calculate the balanced prediction accuracy of a kmeans clustering algorithm which uses the first 2 PLS components from original and imputed measurements:

```{r}
# variates <- mmspls$variates$rna[,1:2]
# labels <- gastru.mae$stage
kmeans_accuracy <- function(variates, labels) {
  names(labels) <- rownames(variates)
  labels <- as.data.frame(labels)
  labels$cell <- rownames(labels)
  kmeans_res <- kmeans(x = variates, centers = length(unique(labels$labels)), iter.max = 3000, nstart = 50)
  labels$cluster <- kmeans_res$cluster
  tbl <- table(labels$labels, labels$cluster)
  stage_cluster <- rep(NA, length(unique(labels$cluster)))
  names(stage_cluster) <- rownames(tbl)
  for (stage in rownames(tbl)) {
    stage_cluster[stage] <- which.max(tbl[stage,])
  }
  ## order columns
  tbl <- tbl[,stage_cluster[rownames(tbl)]]
  total_cells <- rowSums(tbl)
  class_weight <- total_cells / sum(total_cells)
  misclassified <- rowSums(tbl) - diag(tbl)
  err_class <- misclassified / total_cells
  balanced_err <- mean(err_class)
  1 - balanced_err
}
```

Balanced accuracy:

```{r}
set.seed(427)
df <- data.frame(
  ## without imputation
    accuracy_noImpute = sapply(mmspls$variates, function(x) {
        kmeans_accuracy(variates = x[,1:2], labels = mmspls$stage)
    }),
    ## with imputation
    accuracy_impute = sapply(mmspls.imputed$variates, function(x) {
        kmeans_accuracy(variates = x[,1:2], labels = mmspls$stage)
    })
)

kable(round(t(df), 2))
```

The extracted components show stronger phenotypic relevance in terms of embryonic stage for genebody and CGI methylation measurements. This was also apparent from score plots where different stages were better separated for these views after imputation.

# Conclusion

Multi-modal PLS derives the common variation between transcriptome and methylome in different genomic regions for `r dim(gastru.mae@ExperimentList$rna)[2]` single cells during mouse gastrulation. The common axes of variation are driven by the stages of embryo, mainly by variation between early-stage and late-stage cells. The measurements include high levels of masked values and the method ignores these missing elements during integration. Imputation of missing values using a nearest neighbours imputation algorithm enabled the PLS method to capture more cross-modality covariation relevant to embryonic stages.
