<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>scNMTseq challenge analysis using a PLS-based approach â€¢ BIRSBIO2020.scNMTseq.PLS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="scNMTseq challenge analysis using a PLS-based approach">
<meta property="og:description" content="BIRSBIO2020.scNMTseq.PLS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BIRSBIO2020.scNMTseq.PLS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/index.html">scNMTseq challenge analysis using a PLS-based approach</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ajabadi/BIRSBIO2020.scNMTseq.PLS">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="index_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>scNMTseq challenge analysis using a PLS-based approach</h1>
                        <h4 class="author">Al JalalAbadi<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>
</h4>
            
            <h4 class="date">12 Aug 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ajabadi/BIRSBIO2020.scNMTseq.PLS/blob/master/vignettes/index.Rmd"><code>vignettes/index.Rmd</code></a></small>
      <div class="hidden name"><code>index.Rmd</code></div>

    </div>

    
    
<hr>
<p><strong>Note:</strong> These analyses are different from the ones presented in the hackathon due to updated preprocessing if data in order to harmonize the analyses for publication. For original analyses see <a href="https://github.com/ajabadi/scNMT_seq_gastrulation" class="uri">https://github.com/ajabadi/scNMT_seq_gastrulation</a>.</p>
<hr>
<p>Load the required packages:</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">BIRSBIO2020.scNMTseq.PLS</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">MultiAssayExperiment</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">scater</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">scran</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">mixOmics</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggplot2</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">magrittr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">reshape2</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">uwot</span>)</pre></body></html></div>
<div id="data" class="section level1">
<h1 class="hasAnchor">
<a href="#data" class="anchor"></a>Data</h1>
<p>Details of the hackathon data and preprocessing steps: <a href="https://github.com/BIRSBiointegration/Hackathon/tree/master/scNMT-seq" class="uri">https://github.com/BIRSBiointegration/Hackathon/tree/master/scNMT-seq</a></p>
<p>Load the <code>MultiAssayExperiment</code> object from Cloudstor:</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>(<span class="st">'loading data from Cloudstor ...\n'</span>)
<span class="co">#&gt; loading data from Cloudstor ...</span>
<span class="no">gastru.mae_path</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/connections.html">url</a></span>(<span class="st">'https://cloudstor.aarnet.edu.au/plus/s/jsW7nh4YwThw8Q5/download'</span>)</pre></body></html></div>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">gastru.mae</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span>(<span class="no">gastru.mae_path</span>)</pre></body></html></div>
<!-- ```{r, eval=!params$on_my_mac} -->
<!-- cat(sprintf('loading the following assays using SingleCellMultiModal package: %s\n', paste(params$keep_assays, collapse = ', '))) -->
<!-- gastru.mae <- scNMT(dataType = 'mouse_gastrulation', modes = params$keep_assays, dry.run = FALSE, verbose = FALSE) -->
<!-- ``` -->
<div id="filter-mae-object" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-mae-object" class="anchor"></a>Filter MAE object</h2>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="co">## subset RNA expression and DNA methylation modalities</span>
<span class="no">keep_assays</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(<span class="st">"rna|met"</span>,<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="fu">assays</span>(<span class="no">gastru.mae</span>)))
<span class="no">gastru.mae</span> <span class="kw">&lt;-</span> <span class="no">gastru.mae</span>[,,<span class="no">keep_assays</span>]
<span class="co">#&gt; harmonizing input:</span>
<span class="co">#&gt;   removing 4403 sampleMap rows not in names(experiments)</span>
<span class="co">## remove putative extraembryonic cells</span>
<span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span>(<span class="st">'dropping lineages %s, plus the unassigned lineages.\n'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="no">params</span>$<span class="no">drop_lineages</span>, <span class="kw">collapse</span> <span class="kw">=</span> <span class="st">', '</span>)))
<span class="co">#&gt; dropping lineages Primitive_endoderm, Visceral_endoderm, ExE_ectoderm, plus the unassigned lineages.</span>
<span class="no">gastru.mae</span> <span class="kw">&lt;-</span> <span class="no">gastru.mae</span>[,!(<span class="no">gastru.mae</span>$<span class="no">lineage</span> <span class="kw">%in%</span> <span class="no">params</span>$<span class="no">drop_lineages</span> <span class="kw">|</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">gastru.mae</span>$<span class="no">lineage</span>)),]
<span class="co">## keep cells which pass RNA QC</span>
<span class="no">gastru.mae</span> <span class="kw">&lt;-</span> <span class="no">gastru.mae</span>[,<span class="no">gastru.mae</span>$<span class="no">pass_rnaQC</span><span class="kw">==</span><span class="fl">TRUE</span>,]
<span class="co">## Keep full rna SCE for UMAP</span>
<span class="no">rna.sce</span> <span class="kw">&lt;-</span> <span class="no">gastru.mae</span>@<span class="kw">ExperimentList</span>$<span class="no">rna</span>
<span class="co">## keep cells that also pass QC for DNA methylation</span>
<span class="no">gastru.mae</span> <span class="kw">&lt;-</span> <span class="no">gastru.mae</span>[,<span class="no">gastru.mae</span>$<span class="no">pass_metQC</span><span class="kw">==</span><span class="fl">TRUE</span>,]
<span class="co">## rna SCE for cells passing met and rna QC</span>
<span class="no">rna.sce.matching</span> <span class="kw">&lt;-</span> <span class="no">gastru.mae</span>@<span class="kw">ExperimentList</span>$<span class="no">rna</span></pre></body></html></div>
<p>Replace the rna SCE with logcounts in MAE object as the integration wrapper requires matrices in assays:</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="co">## rna is an SCE object</span>
<span class="no">gastru.mae</span>@<span class="kw">ExperimentList</span>
<span class="co">#&gt; ExperimentList class object of length 5:</span>
<span class="co">#&gt;  [1] rna: SingleCellExperiment with 18345 rows and 796 columns</span>
<span class="co">#&gt;  [2] met_genebody: matrix with 17559 rows and 796 columns</span>
<span class="co">#&gt;  [3] met_promoter: matrix with 17179 rows and 796 columns</span>
<span class="co">#&gt;  [4] met_cgi: matrix with 14080 rows and 796 columns</span>
<span class="co">#&gt;  [5] met_DHS: matrix with 6673 rows and 796 columns</span>
<span class="co">## replace SCE with logcounts</span>
<span class="no">gastru.mae</span>@<span class="kw">ExperimentList</span>$<span class="no">rna</span> <span class="kw">&lt;-</span> <span class="fu">logcounts</span>(<span class="no">gastru.mae</span>@<span class="kw">ExperimentList</span>$<span class="no">rna</span>)
<span class="co">## rna updated to matrix of logcounts</span>
<span class="no">gastru.mae</span>@<span class="kw">ExperimentList</span>
<span class="co">#&gt; ExperimentList class object of length 5:</span>
<span class="co">#&gt;  [1] rna: matrix with 18345 rows and 796 columns</span>
<span class="co">#&gt;  [2] met_genebody: matrix with 17559 rows and 796 columns</span>
<span class="co">#&gt;  [3] met_promoter: matrix with 17179 rows and 796 columns</span>
<span class="co">#&gt;  [4] met_cgi: matrix with 14080 rows and 796 columns</span>
<span class="co">#&gt;  [5] met_DHS: matrix with 6673 rows and 796 columns</span></pre></body></html></div>
<p>Overview of assays:</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="co">## helper to get % of missing values in an array</span>
<span class="no">get_pct_missing</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">arr</span>) {
  <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span>(<span class="fl">100</span>*<span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">arr</span>))/<span class="fu"><a href="https://rdrr.io/r/base/prod.html">prod</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">arr</span>)))
}
<span class="co">## get dimensions and % of missing values for all assays</span>
<span class="no">df</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/MultiAssayExperiment/man/MultiAssayExperiment-methods.html">experiments</a></span>(<span class="no">gastru.mae</span>), <span class="kw">function</span>(<span class="no">w</span>)
     {
     <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(
       <span class="st">'N (# cells)'</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">w</span>)[<span class="fl">2</span>],
       <span class="st">'P (# features)'</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">w</span>)[<span class="fl">1</span>],
       <span class="st">'% data missing'</span> <span class="kw">=</span> <span class="fu">get_pct_missing</span>(<span class="no">w</span>)
     )
     })
<span class="no">df</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="no">df</span>)
<span class="fu">kable</span>(<span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="no">df</span>), <span class="kw">align</span> <span class="kw">=</span> <span class="st">'l'</span>)</pre></body></html></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="left">N (# cells)</th>
<th align="left">P (# features)</th>
<th align="left">% data missing</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">rna</td>
<td align="left">796</td>
<td align="left">18345</td>
<td align="left">0</td>
</tr>
<tr class="even">
<td align="left">met_genebody</td>
<td align="left">796</td>
<td align="left">17559</td>
<td align="left">18</td>
</tr>
<tr class="odd">
<td align="left">met_promoter</td>
<td align="left">796</td>
<td align="left">17179</td>
<td align="left">34</td>
</tr>
<tr class="even">
<td align="left">met_cgi</td>
<td align="left">796</td>
<td align="left">14080</td>
<td align="left">45</td>
</tr>
<tr class="odd">
<td align="left">met_DHS</td>
<td align="left">796</td>
<td align="left">6673</td>
<td align="left">59</td>
</tr>
</tbody>
</table>
<p>Breakdown of the number of cells in each stage:</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">gastru.mae</span>$<span class="no">stage</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>() <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span>() <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/magrittr/man/aliases.html">set_rownames</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'stage'</span>, <span class="st">'# of cells'</span>)) <span class="kw">%&gt;%</span> <span class="fu">kable</span>()</pre></body></html></div>
<table class="table"><tbody>
<tr class="odd">
<td align="left">stage</td>
<td align="left">E4.5</td>
<td align="left">E5.5</td>
<td align="left">E6.5</td>
<td align="left">E7.5</td>
</tr>
<tr class="even">
<td align="left"># of cells</td>
<td align="left">62</td>
<td align="left">86</td>
<td align="left">275</td>
<td align="left">373</td>
</tr>
</tbody></table>
</div>
<div id="feature-detection" class="section level2">
<h2 class="hasAnchor">
<a href="#feature-detection" class="anchor"></a>Feature detection</h2>
<p>Create density plots of the feature detection rate across all cells for all modalities:</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="co"># get the methylation assays</span>
<span class="no">met_assays</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">gastru.mae</span>), <span class="kw">pattern</span> <span class="kw">=</span> <span class="st">'^met'</span>, <span class="kw">value</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="co"># add dimensions to labels for ggplot</span>
<span class="no">dims</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/MultiAssayExperiment/man/MultiAssayExperiment-methods.html">experiments</a></span>(<span class="no">gastru.mae</span>[,,<span class="no">met_assays</span>]), <span class="no">dim</span>)
<span class="no">dims</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="no">dims</span>, <span class="kw">function</span>(<span class="no">x</span>) <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span>(<span class="st">' (%s, %s)'</span>, <span class="no">x</span>[<span class="fl">2</span>], <span class="no">x</span>[<span class="fl">1</span>]))
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">met_assays</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="no">met_assays</span>, <span class="no">dims</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">as.list</a></span>()
<span class="co"># calculate the feature detection in a data.frame for methylation assays</span>
<span class="no">coverages</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">met_assays</span>, <span class="kw">function</span>(<span class="no">assay_name</span>) {
  <span class="no">mat</span> <span class="kw">&lt;-</span> <span class="fu">assay</span>(<span class="no">gastru.mae</span>, <span class="no">assay_name</span>)
  <span class="no">NAs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(!<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">mat</span>))/<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">mat</span>)[<span class="fl">2</span>]*<span class="fl">100</span>
  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">pct_NAs</span><span class="kw">=</span><span class="no">NAs</span>)
})
<span class="co"># create a long data.frame containing the assay name for plot</span>
<span class="no">coverages</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://github.com/ajabadi/BIRSBIO2020.scNMTseq.PLS/reference/utils.html">rbindListWithNames</a></span>(<span class="no">coverages</span>)
<span class="no">coverages</span>$<span class="no">dataset</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">coverages</span>$<span class="no">dataset</span>, <span class="kw">levels</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">coverages</span>$<span class="no">dataset</span>), <span class="kw">ordered</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<p>Genomic contexts vary in level of feature detection rates:</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="no">cov_plot</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">coverages</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">pct_NAs</span>)) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span>(<span class="kw">fill</span> <span class="kw">=</span> <span class="st">'lightblue'</span>, <span class="kw">show.legend</span> <span class="kw">=</span> <span class="fl">FALSE</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">xintercept</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(<span class="no">pct_NAs</span>)),
             <span class="kw">color</span><span class="kw">=</span><span class="st">"blue"</span>, <span class="kw">linetype</span><span class="kw">=</span><span class="st">"dashed"</span>, <span class="kw">size</span><span class="kw">=</span><span class="fl">0.5</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="st">'% of cells detecting the feature'</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(<span class="no">.</span>~<span class="no">dataset</span>, <span class="kw">nrow</span> <span class="kw">=</span> <span class="fl">2</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="kw">strip.text.x</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="kw">size</span> <span class="kw">=</span> <span class="fl">10</span>, <span class="kw">face</span> <span class="kw">=</span> <span class="st">'bold'</span>, <span class="kw">color</span> <span class="kw">=</span> <span class="st">'purple'</span>))

<span class="no">cov_plot</span></pre></body></html></div>
<p><img src="index_files/figure-html/unnamed-chunk-12-1.png" width="768"> Density plots for methylation data show that shorter genomic regions tend to have less feature coverage. Dashed blue line indicates the average across all modalities.</p>
</div>
</div>
<div id="rna" class="section level1">
<h1 class="hasAnchor">
<a href="#rna" class="anchor"></a>RNA</h1>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="co"># Create colour palettes for stages:</span>
<span class="no">stages</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"E4.5"</span>, <span class="st">"E5.5"</span>, <span class="st">"E6.5"</span>, <span class="st">"E7.5"</span>)
<span class="co"># col_palette &lt;- dput(viridisLite::viridis(n = length(stages)))</span>
<span class="no">col_palette</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#440154FF"</span>, <span class="st">"#31688EFF"</span>, <span class="st">"#35B779FF"</span>, <span class="st">"#FDE725FF"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">col_palette</span>) <span class="kw">&lt;-</span> <span class="no">stages</span></pre></body></html></div>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="co">## helper function to create UMAP plots coloured by stage</span>
<span class="no">plot_reducedDims</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">sce</span>, <span class="no">reducedDim</span> <span class="kw">=</span> <span class="st">'UMAP'</span>,
                             <span class="no">dims</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">2</span>), <span class="no">col_palette</span>) {
  <span class="no">comps.prefix</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="no">reducedDim</span> <span class="kw">==</span> <span class="st">'UMAP'</span>, <span class="st">'UMAP_'</span>, <span class="st">'PC_'</span>)
  <span class="no">df</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="fu">reducedDim</span>(<span class="no">sce</span>, <span class="no">reducedDim</span>))[,<span class="no">dims</span>]
  <span class="no">axes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">df</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="no">comps.prefix</span>, <span class="no">dims</span>)
  <span class="no">df</span>$<span class="no">stage</span> <span class="kw">&lt;-</span> <span class="no">sce</span>$<span class="no">stage</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes_.html">aes_string</a></span>(<span class="no">axes</span>[<span class="fl">1</span>], <span class="no">axes</span>[<span class="fl">2</span>])) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">col</span><span class="kw">=</span><span class="no">stage</span>), <span class="kw">alpha</span> <span class="kw">=</span> <span class="no">params</span>$<span class="no">ggplot_alpha</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span>()+ <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span>(<span class="kw">values</span> <span class="kw">=</span> <span class="no">col_palette</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">col</span> <span class="kw">=</span> <span class="st">'Stage'</span>)
}</pre></body></html></div>
<div id="dimension-reduction-all-of-rna-cells" class="section level2">
<h2 class="hasAnchor">
<a href="#dimension-reduction-all-of-rna-cells" class="anchor"></a>Dimension reduction all of RNA cells</h2>
<p>Select highly variable genes after accounting for technical variation:</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="no">decomp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html">modelGeneVar</a></span>(<span class="no">rna.sce</span>)
<span class="co">## filter by mean expression and significance of biological variation signal</span>
<span class="no">hvgs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">decomp</span>)[<span class="no">decomp</span>$<span class="no">p.value</span><span class="kw">&lt;</span><span class="fl">0.01</span> <span class="kw">&amp;</span> <span class="no">decomp</span>$<span class="no">mean</span> <span class="kw">&gt;</span> <span class="fl">0.01</span>]
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">hvgs</span>)
<span class="co">#&gt; [1] 322</span></pre></body></html></div>
<p>Run PCA and then UMAP using PCs:</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">npcs</span> <span class="kw">&lt;-</span> <span class="fl">15</span>
<span class="co">## PCA first: retrieve npcs PCs</span>
<span class="no">rna.sce</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runPCA.html">runPCA</a></span>(<span class="no">rna.sce</span>,  <span class="kw">ncomponents</span> <span class="kw">=</span> <span class="no">npcs</span>, <span class="kw">subset_row</span><span class="kw">=</span><span class="no">hvgs</span>, <span class="kw">name</span> <span class="kw">=</span> <span class="st">'PCA'</span>)

<span class="co">## UMAP parameters used:</span>
<span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span>(<span class="st">'Running UMAP with parameters %s\n'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">params</span>$<span class="no">umap_params</span>), <span class="st">':'</span>,<span class="no">params</span>$<span class="no">umap_params</span>, <span class="kw">collapse</span> <span class="kw">=</span> <span class="st">', '</span>)))
<span class="co">#&gt; Running UMAP with parameters run.seed : 42, n_neighbors : 15, n_components : 2, min_dist : 0.55</span>

<span class="co">## run UMAP</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="no">params</span>$<span class="no">umap_params</span>[<span class="st">'run.seed'</span>])
<span class="no">rna.sce</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html">runUMAP</a></span>(<span class="no">rna.sce</span>, <span class="kw">dimred</span><span class="kw">=</span><span class="st">"PCA"</span>,
                   <span class="kw">ncomponents</span> <span class="kw">=</span> <span class="no">params</span>$<span class="no">umap_params</span>[<span class="st">'n_components'</span>],
                   <span class="kw">n_neighbors</span> <span class="kw">=</span> <span class="no">params</span>$<span class="no">umap_params</span>[<span class="st">'n_neighbors'</span>],
                   <span class="kw">min_dist</span> <span class="kw">=</span> <span class="no">params</span>$<span class="no">umap_params</span>[<span class="st">'min_dist'</span>])</pre></body></html></div>
<div id="umap" class="section level3">
<h3 class="hasAnchor">
<a href="#umap" class="anchor"></a>UMAP</h3>
<p>Plot the first two UMAP components:</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="fu">plot_reducedDims</span>(<span class="kw">sce</span> <span class="kw">=</span> <span class="no">rna.sce</span>, <span class="kw">reducedDim</span> <span class="kw">=</span> <span class="st">'UMAP'</span>, <span class="kw">dims</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">2</span>), <span class="kw">col_palette</span> <span class="kw">=</span> <span class="no">col_palette</span>)</pre></body></html></div>
<p><img src="index_files/figure-html/unnamed-chunk-18-1.png" width="768"></p>
<p>UMAP plot separates the cell populations from early-stage, mid-stage, and late-stage cells and captures the the within-stage variation for early-stage and mid-stage cells.</p>
</div>
</div>
<div id="dimension-reduction-of-all-matching-rna-cells-used-in-integration" class="section level2">
<h2 class="hasAnchor">
<a href="#dimension-reduction-of-all-matching-rna-cells-used-in-integration" class="anchor"></a>Dimension reduction of all matching RNA cells used in integration</h2>
<p>Select highly variable genes after accounting for technical variation:</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="no">decomp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html">modelGeneVar</a></span>(<span class="no">rna.sce.matching</span>)
<span class="co">## filter by mean expression and significance of biological variation signal</span>
<span class="no">hvgs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">decomp</span>)[<span class="no">decomp</span>$<span class="no">p.value</span><span class="kw">&lt;</span><span class="fl">0.01</span> <span class="kw">&amp;</span> <span class="no">decomp</span>$<span class="no">mean</span> <span class="kw">&gt;</span> <span class="fl">0.01</span>]
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">hvgs</span>)
<span class="co">#&gt; [1] 347</span></pre></body></html></div>
<p>Run PCA and then UMAP using PCs:</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">npcs</span> <span class="kw">&lt;-</span> <span class="fl">15</span>
<span class="co">## PCA first: retrieve npcs PCs</span>
<span class="no">rna.sce.matching</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runPCA.html">runPCA</a></span>(<span class="no">rna.sce.matching</span>, <span class="kw">ncomponents</span> <span class="kw">=</span> <span class="no">npcs</span>, <span class="kw">subset_row</span><span class="kw">=</span><span class="no">hvgs</span>, <span class="kw">name</span> <span class="kw">=</span> <span class="st">'PCA'</span>)

<span class="co">## UMAP parameters used:</span>
<span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span>(<span class="st">'Running UMAP with parameters %s\n'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">params</span>$<span class="no">umap_params</span>), <span class="st">':'</span>,<span class="no">params</span>$<span class="no">umap_params</span>, <span class="kw">collapse</span> <span class="kw">=</span> <span class="st">', '</span>)))
<span class="co">#&gt; Running UMAP with parameters run.seed : 42, n_neighbors : 15, n_components : 2, min_dist : 0.55</span>

<span class="co">## run UMAP</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="no">params</span>$<span class="no">umap_params</span>[<span class="st">'run.seed'</span>])
<span class="no">rna.sce.matching</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html">runUMAP</a></span>(<span class="no">rna.sce.matching</span>, <span class="kw">dimred</span><span class="kw">=</span><span class="st">"PCA"</span>,
                   <span class="kw">ncomponents</span> <span class="kw">=</span> <span class="no">params</span>$<span class="no">umap_params</span>[<span class="st">'n_components'</span>],
                   <span class="kw">n_neighbors</span> <span class="kw">=</span> <span class="no">params</span>$<span class="no">umap_params</span>[<span class="st">'n_neighbors'</span>],
                   <span class="kw">min_dist</span> <span class="kw">=</span> <span class="no">params</span>$<span class="no">umap_params</span>[<span class="st">'min_dist'</span>])</pre></body></html></div>
<div id="pca" class="section level3">
<h3 class="hasAnchor">
<a href="#pca" class="anchor"></a>PCA</h3>
<p>Plot the first two UMAP components:</p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="fu">plot_reducedDims</span>(<span class="kw">sce</span> <span class="kw">=</span> <span class="no">rna.sce.matching</span>, <span class="kw">reducedDim</span> <span class="kw">=</span> <span class="st">'PCA'</span>, <span class="kw">dims</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">2</span>), <span class="kw">col_palette</span> <span class="kw">=</span> <span class="no">col_palette</span>)</pre></body></html></div>
<p><img src="index_files/figure-html/unnamed-chunk-21-1.png" width="768"></p>
</div>
<div id="umap-1" class="section level3">
<h3 class="hasAnchor">
<a href="#umap-1" class="anchor"></a>UMAP</h3>
<p>Plot the first two UMAP components:</p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="fu">plot_reducedDims</span>(<span class="kw">sce</span> <span class="kw">=</span> <span class="no">rna.sce.matching</span>, <span class="kw">reducedDim</span> <span class="kw">=</span> <span class="st">'UMAP'</span>, <span class="kw">dims</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">2</span>), <span class="kw">col_palette</span> <span class="kw">=</span> <span class="no">col_palette</span>)</pre></body></html></div>
<p><img src="index_files/figure-html/unnamed-chunk-22-1.png" width="768"></p>
<p>Similar results to the analysis with all cells.</p>
</div>
</div>
</div>
<div id="integration" class="section level1">
<h1 class="hasAnchor">
<a href="#integration" class="anchor"></a>Integration</h1>
<!-- ## gene id to gene symbol conversion data.frame -->
<!-- Here we create a reference data.frame to convert Ensembl gene ids to gene symbols for GSEA. -->
<!-- Get all the gene names from all modalities: -->
<!-- ```{r} -->
<!-- all_features <- sapply(experiments(gastru.mae), rownames) %>% unlist() -->
<!-- all_genes <- grep(pattern = '^ENSMUSG', x = all_features, value = TRUE) -->
<!-- all_genes <- unique(all_genes) -->
<!-- length(all_genes) -->
<!-- ``` -->
<!-- Get the Ensembl gene name for these gene ids: -->
<!-- ```{r} -->
<!-- ensembl <- useMart("ensembl", dataset="mmusculus_gene_ensembl") -->
<!-- mouse_gene_ids  <- all_genes -->
<!-- symbols <- getBM(attributes=c('ensembl_gene_id', -->
<!--                               'external_gene_name'), -->
<!--                  filters = "ensembl_gene_id", -->
<!--                  values = all_genes, -->
<!--                  mart = ensembl) -->
<!-- symbols <- data.frame(symbols, row.names = 1) -->
<!-- ``` -->
<!-- ```{r, eval=params$save_output, echo=FALSE} -->
<!-- save(symbols, file = 'savedata/all-gene-symbols.RData') -->
<!-- # load('savedata/all-gene-symbols.RData') -->
<!-- ``` -->
<div id="pls" class="section level2">
<h2 class="hasAnchor">
<a href="#pls" class="anchor"></a>PLS</h2>
<p>data dimensions overview:</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="co"># data dimensions</span>
<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/MultiAssayExperiment/man/MultiAssayExperiment-methods.html">experiments</a></span>(<span class="no">gastru.mae</span>), <span class="no">dim</span>)
<span class="co">#&gt; $rna</span>
<span class="co">#&gt; [1] 18345   796</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $met_genebody</span>
<span class="co">#&gt; [1] 17559   796</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $met_promoter</span>
<span class="co">#&gt; [1] 17179   796</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $met_cgi</span>
<span class="co">#&gt; [1] 14080   796</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $met_DHS</span>
<span class="co">#&gt; [1] 6673  796</span></pre></body></html></div>
<p>integration of all modalities with feature selection:</p>
<!-- ```{r, eval=params$save_output, echo=params$save_output} -->
<!-- pls_output_name <- sprintf('multimodal_spls-ncomp_%s-scale_.rds', ncomp, scale) -->
<!-- ``` -->
<p>Here we select for 50 features on each component in <code>multimodal_sPLS_wrapper</code>:</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="co">## number of components</span>
<span class="no">ncomp</span> <span class="kw">&lt;-</span> <span class="fl">5</span>
<span class="co">## feature scaling</span>
<span class="no">scale</span> <span class="kw">&lt;-</span> <span class="fl">TRUE</span></pre></body></html></div>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span>(<span class="st">'Running PLS with %s components performing variable selection\n'</span>, <span class="no">ncomp</span>))
<span class="co">#&gt; Running PLS with 5 components performing variable selection</span>
<span class="no">st</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>({
  <span class="no">mmspls</span> <span class="kw">&lt;-</span>
    <span class="fu"><a href="https://github.com/ajabadi/BIRSBIO2020.scNMTseq.PLS/reference/multimodal_sPLS_wrapper.html">multimodal_sPLS_wrapper</a></span>(<span class="kw">mae</span> <span class="kw">=</span> <span class="no">gastru.mae</span>, <span class="kw">study_assays</span> <span class="kw">=</span> <span class="kw">NULL</span>,
                            <span class="kw">ncomp</span> <span class="kw">=</span> <span class="no">ncomp</span>, <span class="kw">scale</span> <span class="kw">=</span> <span class="no">scale</span>, <span class="kw">design</span> <span class="kw">=</span> <span class="st">'null'</span>, <span class="kw">lineages</span> <span class="kw">=</span> <span class="kw">NULL</span>,
                            <span class="kw">stages</span> <span class="kw">=</span> <span class="kw">NULL</span>, <span class="kw">DA</span> <span class="kw">=</span> <span class="kw">NULL</span>, <span class="kw">keepX</span> <span class="kw">=</span> <span class="kw">NULL</span>, <span class="kw">save</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
})[<span class="st">'elapsed'</span>]
<span class="co">#&gt; Making unique feature names by adding assay name to feature name for: met_genebody</span>
<span class="co">#&gt; Making unique feature names by adding assay name to feature name for: met_promoter</span>

<span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>(<span class="st">'\nPLS run finished. Runtime: '</span>, <span class="fu">format_runtime</span>(<span class="no">st</span>), <span class="st">'\n'</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt; PLS run finished. Runtime:  00:03:54</span></pre></body></html></div>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="no">plot_pls</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">pls_obj</span>, <span class="no">stage</span>, <span class="no">legend.title</span> <span class="kw">=</span> <span class="st">'Stage'</span>, <span class="no">comp</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">2</span>), <span class="no">col_palette</span>) {
  <span class="no">variates</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">pls_obj</span>$<span class="no">variates</span>, <span class="kw">function</span>(<span class="no">arr</span>){
    <span class="no">df</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="no">arr</span>[,<span class="no">comp</span>])
    <span class="no">df</span>$<span class="no">stage</span> <span class="kw">&lt;-</span> <span class="no">stage</span>
    <span class="no">df</span>
  })
  <span class="no">variates</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://github.com/ajabadi/BIRSBIO2020.scNMTseq.PLS/reference/utils.html">rbindListWithNames</a></span>(<span class="no">variates</span>, <span class="kw">new_col</span> <span class="kw">=</span> <span class="st">'Modality'</span>)
  <span class="no">axes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">variates</span>)[<span class="fl">1</span>:<span class="fl">2</span>]
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">variates</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes_.html">aes_string</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">axes</span>[<span class="fl">1</span>], <span class="kw">y</span> <span class="kw">=</span> <span class="no">axes</span>[<span class="fl">2</span>], <span class="kw">col</span> <span class="kw">=</span> <span class="st">'stage'</span>), <span class="kw">alpha</span> <span class="kw">=</span> <span class="no">params</span>$<span class="no">ggplot_alph</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(<span class="no">.</span>~<span class="no">Modality</span>, <span class="kw">ncol</span> <span class="kw">=</span> <span class="fl">2</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span>(<span class="kw">values</span> <span class="kw">=</span> <span class="no">col_palette</span>)
}</pre></body></html></div>
</div>
<div id="score-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#score-plots" class="anchor"></a>score plots</h2>
<p>Colored by stage:</p>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="fu">plot_pls</span>(<span class="kw">pls_obj</span> <span class="kw">=</span> <span class="no">mmspls</span>, <span class="kw">stage</span> <span class="kw">=</span> <span class="no">gastru.mae</span>$<span class="no">stage</span>, <span class="kw">col_palette</span> <span class="kw">=</span> <span class="no">col_palette</span>)</pre></body></html></div>
<p><img src="index_files/figure-html/unnamed-chunk-28-1.png" width="768"></p>
<p>The derived latent components which seek to maximise the sum of concordant variation b/w rna and other modalities show phenotypic relevance in terms of the embryonic stages. It indicates coordinated changes across the methylomes associated with the distinct transcriptional states of cells in the regions of study during gastrulation.</p>
</div>
</div>
<div id="concordance-with-rna" class="section level1">
<h1 class="hasAnchor">
<a href="#concordance-with-rna" class="anchor"></a>concordance with RNA</h1>
<p>Here we calculate the correlations between components from gene expression and different methylome modalities:</p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="co">## For a pls object which includes pls cell embeddings from a computed latent space</span>
<span class="co">## where covariance with rna is maximal, calculate the correlation b/w rna embeddings</span>
<span class="co">## and those of a given modality's (per component and average) as a measure of concordant variation with rna</span>
<span class="no">concordance</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">pls_obj</span>, <span class="no">comp</span> <span class="kw">=</span> <span class="fl">1</span>) {

  <span class="co">## correlation of components with RNA components</span>
  <span class="no">cor_with_y</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="no">pls_obj</span>$<span class="no">variates</span>[-<span class="fl">1</span>], <span class="kw">function</span>(<span class="no">x_variates</span>) {
    <span class="no">y_variates</span> <span class="kw">&lt;-</span> <span class="no">pls_obj</span>$<span class="no">variates</span><span class="kw">[[</span><span class="fl">1</span>]]
    <span class="no">y_variates</span> <span class="kw">&lt;-</span> <span class="no">y_variates</span>[, <span class="no">comp</span>, <span class="kw">drop</span> <span class="kw">=</span> <span class="fl">FALSE</span>]
    <span class="no">cor</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span>(<span class="no">y_variates</span>, <span class="no">x_variates</span>)
    <span class="no">cor</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span>(<span class="no">cor</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">cor</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">'Component '</span>, <span class="no">comp</span>)
    <span class="no">cor</span>[<span class="st">'mean'</span>] <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(<span class="no">cor</span>)
    <span class="no">cor</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span>(<span class="no">cor</span>, <span class="fl">2</span>)
    <span class="no">cor</span>
  })
  <span class="fu"><a href="https://rdrr.io/pkg/mixOmics/man/S3methods-print.html">print</a></span>(<span class="fu">kable</span>(<span class="no">cor_with_y</span>))
  <span class="co">## make a long data.frame</span>
  <span class="no">cor_with_y</span> <span class="kw">&lt;-</span> <span class="no">cor_with_y</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>() <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span>() <span class="kw">%&gt;%</span>
    <span class="kw pkg">reshape2</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html">melt</a></span>(<span class="kw">id.vars</span> <span class="kw">=</span> <span class="fl">1</span>:<span class="fl">3</span>) <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/magrittr/man/aliases.html">set_colnames</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'Modality'</span>, <span class="st">'Component'</span>, <span class="st">'Concordance'</span>))
  <span class="co">## print concordance measures for each componet and the average:</span>
  <span class="co"># sapply(unique(as.character(cor_with_y$Component)), function(z){</span>
  <span class="co">#   cat('#&gt; Component:', z )</span>
  <span class="co">#   dplyr::filter(cor_with_y, Component == z)[,c('Modality', 'Concordance')] %&gt;% </span>
  <span class="co">#     kable() </span>
  <span class="co"># })</span>

  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">cor_with_y</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">Modality</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="no">Concordance</span>)) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_col</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">fill</span> <span class="kw">=</span> <span class="no">Component</span>), <span class="kw">position</span> <span class="kw">=</span> <span class="st">'dodge2'</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">y</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span>(<span class="st">'Correlation with RNA (comp %s) '</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="no">comp</span>, <span class="kw">collapse</span> <span class="kw">=</span> <span class="st">', '</span>)), <span class="kw">x</span> <span class="kw">=</span> <span class="st">''</span>, <span class="kw">fill</span> <span class="kw">=</span> <span class="st">''</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">ylim</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fl">1</span>)) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="kw">axis.text.x</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="kw">angle</span> <span class="kw">=</span> <span class="fl">45</span>, <span class="kw">vjust</span> <span class="kw">=</span> <span class="fl">0.8</span>, <span class="kw">hjust</span><span class="kw">=</span><span class="fl">1</span>))
}</pre></body></html></div>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="fu">concordance</span>(<span class="kw">pls_obj</span> <span class="kw">=</span> <span class="no">mmspls</span>, <span class="kw">comp</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">2</span>))
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; |            | met_genebody| met_promoter| met_cgi| met_DHS|</span>
<span class="co">#&gt; |:-----------|------------:|------------:|-------:|-------:|</span>
<span class="co">#&gt; |Component 1 |         0.95|         0.92|    0.91|    0.92|</span>
<span class="co">#&gt; |Component 2 |         0.46|         0.59|    0.49|    0.39|</span>
<span class="co">#&gt; |mean        |         0.71|         0.75|    0.70|    0.66|</span></pre></body></html></div>
<p><img src="index_files/figure-html/unnamed-chunk-31-1.png" width="768"></p>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>Melbourne Integrative Genomics, The University of Melbourne (<a href="mailto:al.jal.abadi@gmail.com" class="email">al.jal.abadi@gmail.com</a>)<a href="#fnref1" class="footnote-back">â†©ï¸Ž</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Al JalalAbadi.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
